File: src/api/endpoints/send_establish_introduction.rs
use crate::{CircuitId, Logger, RelayId, User, UserId};
use actix_web::{post, web, HttpResponse, Responder};
use anyhow::{Context, Result};
use serde::Deserialize;
use std::sync::Arc;
use tokio::sync::Mutex;

#[derive(Deserialize)]
pub struct SendEstablishIntroductionBody {
    pub relay_id: RelayId,
    pub circuit_id: CircuitId,
}

#[post("/users/{user_id}/send_establish_introduction")]
pub async fn send_establish_introduction(
    data: web::Data<Arc<Mutex<Vec<User>>>>,
    user_id: web::Path<UserId>,
    body: web::Json<SendEstablishIntroductionBody>,
) -> impl Responder {
    let result: Result<()> = async {
        let data_lock = data.lock().await;
        let user = data_lock
            .iter()
            .find(|u| u.user_descriptor.id == *user_id)
            .ok_or_else(|| anyhow::anyhow!("User not found"))?;
        let introduction_id = UserId::new_v4();
        user.send_establish_introduction(body.relay_id, introduction_id, body.circuit_id)
            .context("Failed to send establish introduction")?;
        Ok(())
    }
    .await;

    match result {
        Ok(_) => HttpResponse::Ok().finish(),
        Err(e) => {
            Logger::error(
                "API",
                format!("Error in send_establish_introduction: {}", e),
            );
            HttpResponse::InternalServerError().json(format!("Internal server error: {}", e))
        }
    }
}

File: veilcomm-ui/src/requests/sendEstablishIntroduction.ts
import { toast } from "react-toastify";
import { RelayState, UserState } from "../data";

async function sendEstablishIntroduction(user: UserState, relay: RelayState, circuit_id: string) {
  if (!user || !relay || !circuit_id) {
    toast.error('Please select a user, circuit_id and relay to send establish introduction');
    return;
  }

  try {
    const response = await fetch(`http://127.0.0.1:8081/users/${user.id}/send_establish_introduction`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({
        relay_id: relay.id,
        circuit_id: circuit_id
      }),
    });
    if (response.ok) {
      toast.success('Establish Introduction sent successfully');
    } else {
      toast.error('Failed to send Establish Introduction');
    }
  } catch (error) {
    console.error('Error:', error);
    toast.error('An error occurred while sending Establish Introduction');
  }
};

export default sendEstablishIntroduction;
